### PR TITLE
Add option to enable client authentication

### PR DESCRIPTION
**Pull request by ldejager**
Add option to make use of TLS client authentication using ACM certificates (certificate_authority_arns).

### CODE REVIEW

**Comment by ldejager** [comment]
@angelabad let me know if there is anything else required on the PR.

File: variables.tf [hunk 1]
```diff
@@ -106,3 +106,15 @@ variable "firehose_logs_delivery_stream" {
   type        = string
   default     = ""
 }
+
+variable "enable_client_authentication" {
+  description = "Enable client authentication. Default value: false."
+  type        = string
```
Comment (kgosietsileM) [inline variables.tf]: Hi @ldejager should this not be a bool?

File: main.tf [hunk 2]
```diff
@@ -1,6 +1,9 @@
 locals {
-  server_properties = join("\n", [for k, v in var.server_properties : format("%s = %s", k, v)])
-  enable_logs       = var.s3_logs_bucket != "" || var.cloudwatch_logs_group != "" || var.firehose_logs_delivery_stream != "" ? ["true"] : []
+  server_properties            = join("\n", [for k, v in var.server_properties : format("%s = %s", k, v)])
+  enable_logs                  = var.s3_logs_bucket != "" || var.cloudwatch_logs_group != "" || var.firehose_logs_delivery_stream != "" ? ["true"] : []
+  enable_client_authentication = var.enable_client_authentication
+  client_authentication        = local.enable_client_authentication ? 1 : 0
```
Comment (angelabad) [inline main.tf]: Sorry but I don't understand this code, why if you have a variable rewrite it as local?

Comment (angelabad) [inline main.tf]: I dont understand the logic either, if client authtentication tls only has one parameter (certificate_authority_arns), why not apply it only if variable like client_authentication_tls_certificate_authority_arns   is setted and avoid the use of client_authentication_enabled?

**Review by angelabad — CHANGES_REQUESTED** [review]
Hi @ldejager and @kgosietsileM, the idea is fine but not the approach, you can fix the issue or I will make the change. (mentioning you, of course)
Thanks for your contribution!

File: main.tf (hunk 2 above)
Comment (ldejager) [reply main.tf]: @angelabad please see comments below.

`enable_client_authentication` is made available to the operator to toggle the conditional creation of resources by means of true/false in the variable definition.

`client_authentication` is making that conditional available to the dynamic block for the certificate authority ARNs.

https://www.terraform.io/docs/language/values/locals.html
https://www.terraform.io/docs/language/expressions/index.html
https://www.terraform.io/docs/language/expressions/conditionals.html

Comment (ldejager) [reply main.tf]: The dynamic block is making use of for_each and deliberately set so it;

1. Defaults to `false`
2. Does not require you to do any check if the variable is set as it's done via the locals' definition (centralised if you will)

**Comment by ldejager** [comment]
@angelabad if you're set on having the code look at whether variables are set in this fashion let me know and I'll amend the PR.

```
for_each = var.s3_logs_bucket != "" ? ["true"] : []
```

**Comment by angelabad** [comment]
> for_each = var.s3_logs_bucket != "" ? ["true"] : []

Hi, @ldejager, it sound good to me, you can use this approach to certificate arns.

Thanks!

**Comment by ldejager** [comment]
@angelabad the changes have been made as requested.

**Comment by ldejager** [comment]
@angelabad can you please review and provide feedback.

**Comment by angelabad** [comment]
Hi @ldejager, sorry for my delay, for this feature, nowadays aws provider is buggy, so Im wating for this:

* https://github.com/hashicorp/terraform-provider-aws/issues/20956

Thanks in advance!

**Comment by angelabad** [comment]
This is closed in this pr: https://github.com/angelabad/terraform-aws-msk-cluster/pull/21

Thanks

**Pull request closed without merging**